### ISSUE TITLE
[Question] Paired data with overlapping samples

### ISSUE THREAD

**Issue by agamemnonc** [KevinMusgrave/pytorch-metric-learning#263]
I have a dataset that yields pairs of original samples (not augmented versions of the same sample) and a target label indicating whether these are similar. E.g. with `batch_size=4`:

```python
[[sample_1, sample_2, 1],
[sample_1, sample_3, 1],
[sample_1, sample_4, 0],
[sample_2, sample_5, 1]]
```
How should I shape my inputs to the loss function in this case? 

Looking at the simple [example](https://kevinmusgrave.github.io/pytorch-metric-learning/#using-loss-functions-for-unsupervised-self-supervised-learning) on the homepage, a separate label is assigned to each image (and to its augmented counter-part), but I am assuming this only applies to the case where all the original images have different ground-truth labels? 

What I am really asking is that, if I follow the convention from the example and reshape my data as:

```python
embeddings = [sample_1, sample_1, sample_1, sample_2, sample_2, sample_3, sample_4, sample_5]
labels = [1, 2, 3, 4, 1, 2, 3, 4]
```
wouldn't this lead to undesired outcome, given that `sample_1` would be assigned 3 different labels?

Thanks for the help.

**Comment by KevinMusgrave** [comment]
In your example, I assume 1 means similar.  Since ```sample_1``` is similar to ```sample_2```, do you also want ```sample_1``` to be considered similar to ```sample_5```? In other words do you want the similarity to be transitive?

## Intransitive similarity
```python
embeddings = [sample_1, sample_1, sample_1, sample_2, sample_2, sample_3, sample_4, sample_5]
is_similar = torch.LongTensor([1, 1, 0, 1])

# positive indices
a1 = torch.where(is_similar==1)[0]
p = a1 + len(is_similar)

# negative indices
a2 = torch.where(is_similar==0)[0]
n = a2 + len(is_similar)

# Create dummy labels
# These won't be used, but at the moment, all loss functions require it regardless
labels = torch.zeros(len(embeddings))
loss = loss_fn(embeddings, labels, (a1, p, a2, n))
```

## Transitive similarity

In the following example, the ```unique_identifier``` array contains strings that uniquely identify each embedding.

```python
embeddings = torch.randn(8, 32)
unique_identifier = ["sample_1", "sample_1", "sample_1", "sample_2", "sample_2", "sample_3", "sample_4", "sample_5"]
is_similar = torch.LongTensor([1, 1, 0, 1])

embedding_to_label_dict = {}
for e in unique_identifier:
    embedding_to_label_dict[e] = -1

labels = torch.zeros(len(embeddings)) - 1
curr_label = 0

for i, x in enumerate(is_similar):
    other_index = i + len(is_similar)
    curr_identifier = unique_identifier[i]
    other_identifier = unique_identifier[other_index]
    stored_label = embedding_to_label_dict[curr_identifier]
    if x == 1:
        if stored_label != -1:
            labels[i] = stored_label
            labels[other_index] = stored_label
            embedding_to_label_dict[other_identifier] = stored_label
        else:
            labels[i] = curr_label
            labels[other_index] = curr_label
            embedding_to_label_dict[curr_identifier] = curr_label
            embedding_to_label_dict[other_identifier] = curr_label
            curr_label += 1
    else:
        if stored_label != -1:
            labels[i] = stored_label
            labels[other_index] = curr_label
            embedding_to_label_dict[other_identifier] = curr_label
            curr_label += 1
        else:
            labels[i] = curr_label
            embedding_to_label_dict[curr_identifier] = curr_label
            curr_label += 1
            labels[other_index] = curr_label
            embedding_to_label_dict[other_identifier] = curr_label
            curr_label += 1
            
loss = loss_fn(embeddings, labels)  
```

The above code produces the following labels:

```python
unique_identifier = ["sample_1", "sample_1", "sample_1", "sample_2", "sample_2", "sample_3", "sample_4", "sample_5"]
> labels tensor([0., 0., 0., 0., 0., 0., 1., 0.])

unique_identifier = ["sample_1", "sample_3", "sample_5", "sample_7", "sample_2", "sample_4", "sample_6", "sample_8"]
> labels tensor([0., 1., 2., 4., 0., 1., 3., 4.])
```

Hope that helps

**Comment by agamemnonc** [comment]
@KevinMusgrave Thank you very much for taking the time and effort to provide such a detailed and thorough answer to my questiion, that's much appreciated.

**Comment by KevinMusgrave** [comment]
Starting in v1.5.0, losses like ContrastiveLoss and TripletMarginLoss no longer require dummy labels if `indices_tuple` is passed in:

```python
loss = loss_fn(embeddings, indices_tuple=triplets)
```

(Posting here for future readers.)
